### PR TITLE
Client: New Mechanism to Keep Peer Latest Block Updated

### PR DESCRIPTION
**Pull request by holgerd77**
~This PR adds a periodical check if the ETH status message exchanged by peers in the client is getting too old for then triggering a new handshake which should (theoretically) update the status information.~

~Not yet working, will continue tomorrow. Feedback welcome!~

Initial description outdated, see updated post below

### CODE REVIEW

**Comment by holgerd77** [comment]
Yeah, ok, I think this should basically do the job! 🙂 

Took me several iterations to get this a bit to the point, the base implementation is now pretty clean I would say. So a more differentiated request on the latest/best block by a peer is now directly integrated into the `peer.latest()` method.

The peer remembers the last block from the best block request, on first round this is the block from `STATUS` (to get a first reference point, since there is no block number until then).

On second round onwards, a request is done either towards a block number from `config.syncTargetHeight`, or a forward calculated potential value going from the last best block and taking timestamps and the 12s slot period into account, whichever one is higher.

This logic should work both pre and post merge. I *only* tested this pre-merge on mainnet, this is then actually the first time we have something like a constant sync target height update on mainnet (since the `latest()` call is also used to update this `syncTargetHeight` in case this is higher than the existing one).

Two caveats remain:

1. Atm it is not safe to call this "out of bounds" in a regular interval on non-idle peers in peer pool, since this disturbs Fetcher requests as long as we do not support request IDs (so that ETH requests remain ordered, introduced with eth/66, we implemented only the "dummy" way of just reading this but throwing away). I have "solved" this now by hacking this into the different fetchers directly, within a Fetcher `request()` call, to make sure that ETH calls from `getBlockHeaders()` and from the fetcher calls do not interfer. This already works astonishingly well, also the interval of these calls seems bearable. So from my judgement despite being locally hacky I think we could live with this for practical reasons.

2. Along the work I refactored the location of the `latest()` method from being associated with the `Sync` module to move into the `Peer`. This was getting obvious that this is the right choice, all calling was going into peer already before and now with the per-peer-saving of the best block and stuff it became even more apparent that the method was misplaced in `Sync` (also considering that this update actually has nothing to do with the sync activity of the client). So I would very much like to keep this refactoring. Unfortunately this triggers a couple of tests to fail, mainly in the sync tests. This is solely due to the necessity of some changing in the mocking. Unfortunately I am not such a mocking expert, I seriously tried for 1 1/2 hours but did not get this to work. 😏 I worked my way through [vi.mock()](https://vitest.dev/api/vi.html#mock-modules), but with no success yet (basically `peer.latest()` now needs to be mocked instead of `sync.latest()`. @scorbajio since you have done a lot with mocking, maybe you can also take a look, if you have the ressources? Feel free to directly push to this branch if you do!

File: packages/client/src/net/peer/peer.ts [hunk 1]
```diff
@@ -108,28 +168,25 @@ export abstract class Peer extends EventEmitter {
...
     } else if (protocol.name === 'snap') {
+      bound = new BoundSnapProtocol(boundOpts)
+      if (sender.status === undefined) throw Error('Snap can only be bound on handshaked peer')
+
       this.snap = <BoundSnapProtocol>bound
-    } else if (protocol.name === 'les') {
-      this.les = <BoundLesProtocol>bound
+    } else {
+      throw new Error(`addProtocol: ${protocol.name} protocol not supported`)
     }
```
Comment (holgerd77) [inline packages/client/src/net/peer/peer.ts:188]: This looks wild but is only a small local refactor making things a bit cleaner. 🙂

**Review by g11tech — APPROVED** [review]
lgtm

**Review by jochem-brouwer — COMMENTED** [review]
Got one question, LGTM in general

File: packages/client/src/net/peer/peer.ts [hunk 2]
```diff
@@ -90,6 +93,64 @@ export abstract class Peer extends EventEmitter {
...
+   * header timestamp "forward-calculated" by block/slot times (12s).
+   */
+  getPotentialBestHeaderNum(): bigint {
+    let forwardCalculatedNum = BIGINT_0
+    const bestSyncTargetNum = this.config.syncTargetHeight ?? BIGINT_0
+    if (this.eth?.updatedBestHeader !== undefined) {
+      const bestHeaderNum = this.eth!.updatedBestHeader.number
+      const nowSec = Math.floor(Date.now() / 1000)
+      const diffSec = nowSec - Number(this.eth!.updatedBestHeader.timestamp)
+      const SLOT_TIME = 12
```
[the remark this answers is not in the document]
Comment (holgerd77) [reply packages/client/src/net/peer/peer.ts:146]: Good point, I initially planned to do 2-3 more commits fine tuning the existing solution a bit with a somewhat more adaptive algorithm (basically remember the last unsuccessful try (or remove on success again) and if present then gradually downgrade the request (so: try a lower block number).

Guess we likely can/want to do this in a small follow-up PR, should be not such a big effort.

**Pull request merged**